src/hotspot/share/utilities/debug.cpp

```diff
@@ -76,8 +76,19 @@ static intx g_asserting_thread = 0;
 static void* g_assertion_context = nullptr;
 #endif // CAN_SHOW_REGISTERS_ON_ASSERT
 
-// Set to suppress secondary error reporting.
-bool Debugging = false;
+int DebuggingContext::_enabled = 0; // Initially disabled.
+
+DebuggingContext::DebuggingContext() {
+  _enabled += 1;                // Increase nesting count.
+}
+
+DebuggingContext::~DebuggingContext() {
+  if (is_enabled()) {
+    _enabled -= 1;              // Decrease nesting count.
+  } else {
+    fatal("Debugging nesting confusion");
```

Comment:
Why did you make this change?  Isn't this an RAII object - how would this be a problem?

Reply:
The earlier version (from PR#12845) used `assert(is_enabled(), "...")`.
But that's vacuous, being equivalent to `assert(false, ...)`.  (Yeah, that's
obvious.)

Of course it should never fail, but that's true for a lot of these kinds of
things.   I also thought about just discarding the check entirely.